File: apps/api/drift_api/service.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Dict, Optional

from drift_api.config import ApiConfig
from drift_api.errors import ApiError
from drift_api.gcs import parse_gcs_uri
from drift_api.ids import new_job_id, new_upload_id
from drift_api.repositories import (
    FirestoreJobRepository,
    FirestoreUploadRepository,
    JobRecord,
    UploadRecord,
)
from drift_api.storage import StorageService
from drift_api.tasks import CloudTasksEnqueuer


class DriftApiService:
    def __init__(
        self,
        config: ApiConfig,
        storage_service: StorageService,
        upload_repository: FirestoreUploadRepository,
        job_repository: FirestoreJobRepository,
        task_enqueuer: CloudTasksEnqueuer,
    ) -> None:
        self._config = config
        self._storage = storage_service
        self._uploads = upload_repository
        self._jobs = job_repository
        self._tasks = task_enqueuer

    def create_upload(
        self,
        source_file,
        document_files: Sequence[object],
        project_name: Optional[str] = None,
    ) -> Dict[str, object]:
        upload_id = new_upload_id()
        stored_bundle = self._storage.upload_bundle(
            upload_id,
            source_file,
            document_files,
            max_document_files=self._config.max_document_files,
        )
        record = UploadRecord(
            upload_id=upload_id,
            source_archive_uri=stored_bundle.source.uri,
            document_uris=tuple(document.uri for document in stored_bundle.documents),
            project_name=_optional_text(project_name),
            source_file_name=stored_bundle.source.original_filename,
            document_file_names=tuple(
                document.original_filename for document in stored_bundle.documents
            ),
        )
        self._uploads.create(record)
        return _upload_response(record)

    def create_job(self, raw_body: Mapping[str, object]) -> Dict[str, object]:
        upload_id = _optional_text(_read_first(raw_body, "uploadId", "upload_id"))
        upload_record: Optional[UploadRecord] = None
        if upload_id:
            upload_record = self._uploads.get(upload_id)
            if upload_record is None:
                raise ApiError(404, "upload_not_found", f"Upload not found: {upload_id}")

        source_archive_uri = _optional_text(
            _read_first(raw_body, "sourceArchiveUri", "source_archive_uri")
        )
        document_uris = _read_first(raw_body, "documentUris", "document_uris", "documents")
        project_name = _optional_text(_read_first(raw_body, "projectName", "project_name"))
        requested_by = _optional_text(_read_first(raw_body, "requestedBy", "requested_by"))

        if upload_record is not None:
            source_archive_uri = source_archive_uri or upload_record.source_archive_uri
            document_uris = document_uris if document_uris is not None else upload_record.document_uris
            project_name = project_name or upload_record.project_name

        if not source_archive_uri:
            raise ApiError(400, "missing_source_archive", "sourceArchiveUri or uploadId is required")
        parse_gcs_uri(source_archive_uri, "sourceArchiveUri")

        normalized_document_uris = _normalize_document_uris(document_uris)
        for index, document_uri in enumerate(normalized_document_uris):
            parse_gcs_uri(document_uri, f"documentUris[{index}]")

        job_id = new_job_id()
        results_prefix = self._config.results_prefix_template.format(job_id=job_id).strip("/")
        record = JobRecord(
            job_id=job_id,
            upload_id=upload_id,
            project_name=project_name,
            source_archive_uri=source_archive_uri,
            document_uris=normalized_document_uris,
            results_prefix=results_prefix,
            status="queued",
        )
        self._jobs.create_queued(record)

        task_payload = {
            "jobId": job_id,
            "sourceArchiveUri": source_archive_uri,
            "documentUris": list(normalized_document_uris),
            "resultsPrefix": results_prefix,
        }
        if project_name:
            task_payload["projectName"] = project_name
        if requested_by:
            task_payload["requestedBy"] = requested_by

        try:
            task_name = self._tasks.enqueue_analysis_task(task_payload)
        except Exception as exc:
            self._jobs.mark_failed(job_id, str(exc))
            raise

        response = _job_response(record)
        if task_name:
            response["taskName"] = task_name
        return response

    def get_job(self, job_id: str) -> Dict[str, object]:
        record = self._jobs.get(job_id)
        if record is None:
            raise ApiError(404, "job_not_found", f"Job not found: {job_id}")
        return _job_response(record)

    def get_results(self, job_id: str) -> Dict[str, object]:
        record = self._jobs.get(job_id)
        if record is None:
            raise ApiError(404, "job_not_found", f"Job not found: {job_id}")
        if record.status != "succeeded":
            raise ApiError(
                409,
                "job_not_succeeded",
                "Job results are available only after the job succeeds",
                {"status": record.status},
            )
        if not record.artifact_paths:
            raise ApiError(404, "results_not_found", f"No artifacts found for job: {job_id}")

        artifacts = []
        for name, uri in sorted(record.artifact_paths.items()):
            parse_gcs_uri(uri, f"artifactPaths.{name}")
            artifacts.append(
                {
                    "name": name,
                    "uri": uri,
                    "url": self._storage.signed_url(
                        uri,
                        self._config.signed_url_expiration_seconds,
                    ),
                }
            )

        return {
            "jobId": record.job_id,
            "status": record.status,
            "expiresIn": self._config.signed_url_expiration_seconds,
            "artifacts": artifacts,
        }


def build_service(config: ApiConfig) -> DriftApiService:
    storage_service = StorageService(
        config.uploads_bucket,
        uploads_prefix_template=config.uploads_prefix_template,
    )
    return DriftApiService(
        config=config,
        storage_service=storage_service,
        upload_repository=FirestoreUploadRepository(config.firestore_uploads_collection),
        job_repository=FirestoreJobRepository(config.firestore_jobs_collection),
        task_enqueuer=CloudTasksEnqueuer(config),
    )


def _upload_response(record: UploadRecord) -> Dict[str, object]:
    response: Dict[str, object] = {
        "uploadId": record.upload_id,
        "sourceArchiveUri": record.source_archive_uri,
        "documentUris": list(record.document_uris),
    }
    if record.project_name:
        response["projectName"] = record.project_name
    return response


def _job_response(record: JobRecord) -> Dict[str, object]:
    response: Dict[str, object] = {
        "jobId": record.job_id,
        "status": record.status,
        "sourceArchiveUri": record.source_archive_uri,
        "documentUris": list(record.document_uris),
        "resultsPrefix": record.results_prefix,
    }
    if record.upload_id:
        response["uploadId"] = record.upload_id
    if record.project_name:
        response["projectName"] = record.project_name
    if record.artifact_paths:
        response["artifactPaths"] = record.artifact_paths
    if record.error_message:
        response["errorMessage"] = record.error_message
    return response


def _normalize_document_uris(value: object) -> Sequence[str]:
    if value is None:
        raise ApiError(400, "missing_documents", "documentUris or uploadId is required")
    if not isinstance(value, Sequence) or isinstance(value, (str, bytes)):
        raise ApiError(400, "invalid_documents", "documentUris must be an array")
    if not value:
        raise ApiError(400, "missing_documents", "documentUris must not be empty")
    return tuple(str(item) for item in value)


def _read_first(mapping: Mapping[str, object], *keys: str) -> object:
    for key in keys:
        if key in mapping:
            return mapping[key]
    return None


def _optional_text(value: object) -> Optional[str]:
    if value is None:
        return None
    text = str(value).strip()
    return text or None

File: apps/api/src/service.ts
import { ApiConfig } from "./config";
import { ApiError } from "./errors";
import { parseGcsUri } from "./gcs";
import { newJobId, newUploadId } from "./ids";
import {
  FirestoreJobRepository,
  FirestoreUploadRepository,
  JobRecord,
  JobRepository,
  UploadRecord,
  UploadRepository,
  jobRecord,
  uploadRecord,
} from "./repositories";
import { StorageService, UploadedFile } from "./storage";
import { AnalysisTaskPayload, CloudTasksEnqueuer, TaskEnqueuer } from "./tasks";

export type UploadResponse = {
  uploadId: string;
  sourceArchiveUri: string;
  documentUris: string[];
  projectName?: string;
};

export type JobResponse = {
  jobId: string;
  status: string;
  sourceArchiveUri: string;
  documentUris: string[];
  resultsPrefix: string;
  uploadId?: string;
  projectName?: string;
  artifactPaths?: Record<string, string>;
  errorMessage?: string;
  taskName?: string;
};

export type ResultsResponse = {
  jobId: string;
  status: string;
  expiresIn: number;
  artifacts: Array<{
    name: string;
    uri: string;
    url: string;
  }>;
};

export type StorageBoundary = Pick<StorageService, "uploadBundle" | "signedUrl">;

export class DriftApiService {
  private readonly config: ApiConfig;
  private readonly storage: StorageBoundary;
  private readonly uploads: UploadRepository;
  private readonly jobs: JobRepository;
  private readonly tasks: TaskEnqueuer;

  constructor({
    config,
    storageService,
    uploadRepository,
    jobRepository,
    taskEnqueuer,
  }: {
    config: ApiConfig;
    storageService: StorageBoundary;
    uploadRepository: UploadRepository;
    jobRepository: JobRepository;
    taskEnqueuer: TaskEnqueuer;
  }) {
    this.config = config;
    this.storage = storageService;
    this.uploads = uploadRepository;
    this.jobs = jobRepository;
    this.tasks = taskEnqueuer;
  }

  async createUpload(
    sourceFile: UploadedFile | null,
    documentFiles: readonly UploadedFile[],
    projectName: string | null = null,
  ): Promise<UploadResponse> {
    const uploadId = newUploadId();
    const storedBundle = await this.storage.uploadBundle(
      uploadId,
      sourceFile,
      documentFiles,
      this.config.maxDocumentFiles,
    );
    const record = uploadRecord({
      uploadId,
      sourceArchiveUri: storedBundle.source.uri,
      documentUris: storedBundle.documents.map((document) => document.uri),
      projectName: optionalText(projectName),
      sourceFileName: storedBundle.source.originalFilename,
      documentFileNames: storedBundle.documents.map((document) => document.originalFilename),
    });
    await this.uploads.create(record);
    return uploadResponse(record);
  }

  async createJob(rawBody: Record<string, unknown>): Promise<JobResponse> {
    const uploadId = optionalText(readFirst(rawBody, "uploadId", "upload_id"));
    let uploadRecordValue: UploadRecord | null = null;
    if (uploadId) {
      uploadRecordValue = await this.uploads.get(uploadId);
      if (uploadRecordValue == null) {
        throw new ApiError(404, "upload_not_found", `Upload not found: ${uploadId}`);
      }
    }

    let sourceArchiveUri = optionalText(readFirst(rawBody, "sourceArchiveUri", "source_archive_uri"));
    let documentUris: unknown = readFirst(rawBody, "documentUris", "document_uris", "documents");
    let projectName = optionalText(readFirst(rawBody, "projectName", "project_name"));
    const requestedBy = optionalText(readFirst(rawBody, "requestedBy", "requested_by"));

    if (uploadRecordValue != null) {
      sourceArchiveUri = sourceArchiveUri || uploadRecordValue.sourceArchiveUri;
      documentUris = documentUris == null ? uploadRecordValue.documentUris : documentUris;
      projectName = projectName || uploadRecordValue.projectName;
    }

    if (!sourceArchiveUri) {
      throw new ApiError(400, "missing_source_archive", "sourceArchiveUri or uploadId is required");
    }
    parseGcsUri(sourceArchiveUri, "sourceArchiveUri");

    const normalizedDocumentUris = normalizeDocumentUris(documentUris);
    normalizedDocumentUris.forEach((documentUri, index) => parseGcsUri(documentUri, `documentUris[${index}]`));

    const jobId = newJobId();
    const resultsPrefix = this.config.resultsPrefixTemplate.replaceAll("{job_id}", jobId).replace(/^\/+|\/+$/g, "");
    const record = jobRecord({
      jobId,
      uploadId,
      projectName,
      sourceArchiveUri,
      documentUris: normalizedDocumentUris,
      resultsPrefix,
      status: "queued",
    });
    await this.jobs.createQueued(record);

    const taskPayload: AnalysisTaskPayload = {
      jobId,
      sourceArchiveUri,
      documentUris: [...normalizedDocumentUris],
      resultsPrefix,
    };
    if (projectName) {
      taskPayload.projectName = projectName;
    }
    if (requestedBy) {
      taskPayload.requestedBy = requestedBy;
    }

    let taskName: string | null = null;
    try {
      taskName = await this.tasks.enqueueAnalysisTask(taskPayload);
    } catch (error) {
      await this.jobs.markFailed(jobId, error instanceof Error ? error.message : String(error));
      throw error;
    }

    const response = jobResponse(record);
    if (taskName) {
      response.taskName = taskName;
    }
    return response;
  }

  async getJob(jobId: string): Promise<JobResponse> {
    const record = await this.jobs.get(jobId);
    if (record == null) {
      throw new ApiError(404, "job_not_found", `Job not found: ${jobId}`);
    }
    return jobResponse(record);
  }

  async getResults(jobId: string): Promise<ResultsResponse> {
    const record = await this.jobs.get(jobId);
    if (record == null) {
      throw new ApiError(404, "job_not_found", `Job not found: ${jobId}`);
    }
    if (record.status !== "succeeded") {
      throw new ApiError(
        409,
        "job_not_succeeded",
        "Job results are available only after the job succeeds",
        { status: record.status },
      );
    }
    if (!record.artifactPaths || Object.keys(record.artifactPaths).length === 0) {
      throw new ApiError(404, "results_not_found", `No artifacts found for job: ${jobId}`);
    }

    const artifacts = [];
    for (const [name, uri] of Object.entries(record.artifactPaths).sort(([left], [right]) =>
      left.localeCompare(right),
    )) {
      parseGcsUri(uri, `artifactPaths.${name}`);
      artifacts.push({
        name,
        uri,
        url: await this.storage.signedUrl(uri, this.config.signedUrlExpirationSeconds),
      });
    }

    return {
      jobId: record.jobId,
      status: record.status,
      expiresIn: this.config.signedUrlExpirationSeconds,
      artifacts,
    };
  }
}

export function buildService(config: ApiConfig): DriftApiService {
  const storageService = new StorageService(config.uploadsBucket, {
    uploadsPrefixTemplate: config.uploadsPrefixTemplate,
  });
  return new DriftApiService({
    config,
    storageService,
    uploadRepository: new FirestoreUploadRepository(config.firestoreUploadsCollection),
    jobRepository: new FirestoreJobRepository(config.firestoreJobsCollection),
    taskEnqueuer: new CloudTasksEnqueuer(config),
  });
}

export function uploadResponse(record: UploadRecord): UploadResponse {
  const response: UploadResponse = {
    uploadId: record.uploadId,
    sourceArchiveUri: record.sourceArchiveUri,
    documentUris: [...record.documentUris],
  };
  if (record.projectName) {
    response.projectName = record.projectName;
  }
  return response;
}

export function jobResponse(record: JobRecord): JobResponse {
  const response: JobResponse = {
    jobId: record.jobId,
    status: record.status,
    sourceArchiveUri: record.sourceArchiveUri,
    documentUris: [...record.documentUris],
    resultsPrefix: record.resultsPrefix,
  };
  if (record.uploadId) {
    response.uploadId = record.uploadId;
  }
  if (record.projectName) {
    response.projectName = record.projectName;
  }
  if (record.artifactPaths && Object.keys(record.artifactPaths).length > 0) {
    response.artifactPaths = record.artifactPaths;
  }
  if (record.errorMessage) {
    response.errorMessage = record.errorMessage;
  }
  return response;
}

export function normalizeDocumentUris(value: unknown): string[] {
  if (value == null) {
    throw new ApiError(400, "missing_documents", "documentUris or uploadId is required");
  }
  if (!Array.isArray(value)) {
    throw new ApiError(400, "invalid_documents", "documentUris must be an array");
  }
  if (value.length === 0) {
    throw new ApiError(400, "missing_documents", "documentUris must not be empty");
  }
  return value.map((item) => String(item));
}

function readFirst(mapping: Record<string, unknown>, ...keys: string[]): unknown {
  for (const key of keys) {
    if (Object.prototype.hasOwnProperty.call(mapping, key)) {
      return mapping[key];
    }
  }
  return undefined;
}

export function optionalText(value: unknown): string | null {
  if (value == null) {
    return null;
  }
  const text = String(value).trim();
  return text || null;
}
